src/ORM/Query.php

```diff
@@ -424,7 +424,7 @@ public function clearContain()
      * Used to recursively add contained association column types to
      * the query.
      *
-     * @param \Cake\ORM\Table|\Cake\Datasource\RepositoryInterface $table The table instance to pluck associations from.
```

Comment:
The interface should stay for where it is used (instead of concrete one). Both are in use currently.

Reply:
The method has call to `$table->association()` and `RepositoryInterface` doesn't have an `association()` method, so how is it valid?

Reply:
OK, the classes that use the interface approach will then all be "red". Not sure what the worse code is then :/ I guess we have to live with it either way we chose that something will "fail".

Reply:
>  Bummer that we cannot do faster majors to actually clean up those smells.

I hope that the 4->5 journey isn't filled with as many deprecations and method renames as well. We've churned huge parts of the framework API in order to get stricter typehinting. I hope to not have to do that again :smil:

Reply:
> churned huge parts of the framework API 

The main issue I have is the wrong interface usage everywhere and that you cant fix it BC, thus we carry this baggage always for years till it can be resolved.
But sure, API regarding methods and their interface definition is also a problem for such long release cycles.